lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+    int operandIndex = 0;
+    for (auto arg : oper->getOperands()) {
+      int oldArgIndex = -1;
+      for (int i = 0; i < numArgs; i++) {
+        if (arg == op.getInductionVar(i)) {
+          oldArgIndex = i;
+          break;
+        }
+      }
+      if (oldArgIndex != -1) {
+        Value add = op.getBody()->getArgument(tileOffsets[oldArgIndex]);
+        Value mul;
+        for (int j = tileOffsets[oldArgIndex] + 1;
+             j < tileOffsets[oldArgIndex + 1]; j++) {
+          unsigned cumulativeUpperBound = 1;
+          for (int k = j; k < tileOffsets[oldArgIndex + 1]; k++)
+            cumulativeUpperBound *= op.getStaticUpperBound()[j];
+          Value upperBound = rewriter.create<arith::ConstantIndexOp>(
+              op.getLoc(), cumulativeUpperBound);
+          mul = rewriter.create<arith::MulIOp>(op.getLoc(), b.getIndexType(),
+                                               add, upperBound);
+          add = rewriter.create<arith::AddIOp>(
+              op.getLoc(), b.getIndexType(), mul, op.getBody()->getArgument(j));
+        }
+        oper->setOperand(operandIndex, add);
+      }
+      operandIndex++;
+    }
+  }
+
+  for (int i = 1; i <= 3; i++) {
+    auto definingOp = brgemmOp.getOperand(i).getDefiningOp();
+    if (isa<memref::SubViewOp>(definingOp)) {
+      Value operand = definingOp->getOperand(0);
+
+      auto operandType = operand.getType();
+      auto resultType = dyn_cast<MemRefType>(brgemmOp.getOperand(i).getType());
+      auto reassociationIndex =
+          getReassociationIndices(originalShapes[i - 1], tilingVectors[i - 1]);
+
+      SmallVector<int64_t> shape;
+      for (size_t j = 0; j < tilingVectors[i - 1].size(); j++)
+        shape.append(tilingVectors[i - 1][j].begin(),
+                     tilingVectors[i - 1][j].end());
+      shape.append(
+          std::next(originalShapes[i - 1].begin(), tilingVectors[i - 1].size()),
+          originalShapes[i - 1].end());
+      int inductionVarBase = i - 1;
+      if (definingOp == brgemmOp.getOperand(3).getDefiningOp())
+        inductionVarBase = 0;
+      insertSubview(shape, operandType, resultType, reassociationIndex, operand,
+                    op, rewriter, definingOp, inductionVarBase);
+    }
+  }
+
+  return success();
+}
+
+bool getInnermostForLoops(Operation *rootOp,
+                          SmallVectorImpl<scf::ForallOp> &result) {
+  assert(rootOp != nullptr && "Root operation must not be a nullptr.");
+  bool rootEnclosesForAllloops = false;
+  for (Region &region : rootOp->getRegions()) {
+    for (Block &block : region.getBlocks()) {
+      for (Operation &op : block) {
+        bool enclosesPloops = getInnermostForLoops(&op, result);
+        rootEnclosesForAllloops |= enclosesPloops;
+        if (auto ploop = dyn_cast<scf::ForallOp>(op)) {
+          rootEnclosesForAllloops = true;
+
+          // Collect forall loop if it is an innermost one.
+          if (!enclosesPloops)
```

Comment:
This will return true for more than one inner loop at different nest levels, I'm not sure that's what you want.

I'd focus on the parallel loops at the function level and not recurse into further next levels for now.

Reply:
This will return true for the innermost level of a nested forall loop, for each loop nest.